### PR TITLE
Presample

### PR DESCRIPTION
**Pull request by dblenkus**

### CODE REVIEW

File: resolwe_bio/models.py [hunk 1]
```diff
@@ -20,3 +20,6 @@ class Meta(BaseCollection.Meta):
 
     #: list of collections to which sample belong
     collections = models.ManyToManyField(Collection)
+
+    #: identifies is sample is not annotated (``False`` = annotated)
```
Comment (mstajdohar) [inline resolwe_bio/models.py]: Not just that. Here are my notes from a discussion in Slovenian language:

> Sample bo postal sample iz presample, ko:
> - bo imel veljavno anotacijo
> - bo imel podatke
> - bo uporabnik potrdil anotacijo (Klik na gumb finalize/complete annotation) (edited)
> 
> zadnji dve tocki bomo dodali v bodoce
> 
> za zdaj torej ne bo veliko razlike, nam bo pa ta novi endpoint omogocal to razsiritev, pa tudi bolj clear bo, da dokler je nekaj presample, se ne sme uporabljati drugot kot za anotacijo
>

File: resolwe_bio/serializers.py [hunk 2]
```diff
@@ -15,4 +15,12 @@ class SampleSerializer(CollectionSerializer):
...
     class Meta(CollectionSerializer.Meta):
+        fields = CollectionSerializer.Meta.fields + ('collections', 'presample')
+
+
+class PresampleSerializer(CollectionSerializer):
+    collections = serializers.PrimaryKeyRelatedField(many=True, read_only=True)
+
+    class Meta(CollectionSerializer.Meta):
         fields = CollectionSerializer.Meta.fields + ('collections',)
+        read_only_fields = CollectionSerializer.Meta.read_only_fields + ('presample',)
```
Comment (mstajdohar) [inline resolwe_bio/serializers.py]: How can a user set `presample=True` if read only? The idea is that user will confirm (submit/finalize) the annotations when he is happy with it. Three requirement should be satisfied to graduate a presample to sample:
- Annotation is valid
- Data is attached to the sample object
- User is OK with the annotation

File: resolwe_bio/views.py [hunk 3]
```diff
@@ -10,28 +10,33 @@
...
-                queryset = queryset.filter(~Q(descriptor__has_key='geo') | ~Q(descriptor__geo__has_key='annotator'))
-            else:
-                queryset = queryset.filter(Q(descriptor__has_key='geo') & Q(descriptor__geo__has_key='annotator'))
+class SampleViewSet(CollectionViewSet):
+    filter_class = SampleFilter
+    serializer_class = SampleSerializer
 
-        return queryset.order_by('-latest_date')
+    queryset = Sample.objects.annotate(
+        latest_date=Max('data__created')
```
Comment (mstajdohar) [inline resolwe_bio/views.py]: We wish to sort sample by `data__modified` by default. The presamples should be sorted by `data__created` as above.

File: resolwe_bio/tests/unit/test_presample.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        sample.presample = False  # mark sample as annotated
+        sample.save()
+        self.assertEqual(self.sample_queryset.count(), 1)
+        self.assertEqual(self.presample_queryset.count(), 0)
+
+    def test_update(self):
+        sample = Sample.objects.create(contributor=self.user)
+
+        request = self.factory.patch(self.presample_view, data={}, format='json')
```
Comment (mstajdohar) [inline resolwe_bio/tests/unit/test_presample.py]: Test that you can save the `presample=False` and graduate a presample to a sample.

**Comment by mstajdohar** [comment]
Why do the tests fail?

File: resolwe_bio/serializers.py (hunk 2 above)
Comment (dblenkus) [inline resolwe_bio/serializers.py]: +1, I've mixed names (``PresampleSerializer`` and ``SampleSerializer``)

File: resolwe_bio/tests/unit/test_presample.py (hunk 4 above)
Comment (dblenkus) [inline resolwe_bio/tests/unit/test_presample.py]: I'm working on this test, this is last part that is missing. I think that requests will be ok, as presample is still part of the queryset at the moment that request is made.

**Comment by dblenkus** [comment]
Tests fail because https://github.com/genialis/resolwe/pull/87 is required to support changes.

File: resolwe_bio/models.py (hunk 1 above)
Comment (mstajdohar) [inline resolwe_bio/models.py]: Write:
```
#: sample is finalized (has data, annotations and is confirmed)
```

File: docs/CHANGELOG.rst [hunk 5]
```diff
@@ -21,6 +21,7 @@ Changed
 * Updated Cuffquant/Cuffnorm tools
 * Change ROSE2 enhancer rank plot labels
 * Refactor processor syntax
+* Move processes tests into ``processes`` subfolder
```
Comment (tjanez) [inline docs/CHANGELOG.rst]: Please, change 'subfolder' to 'subdirectory' (also in the commit message).

File: resolwe_bio/tests/processes/__init__.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+"""
+**************
+Procesor Tests
```
Comment (tjanez) [inline resolwe_bio/tests/processes/__init__.py]: Should we change this to *Processes Tests*?

File: resolwe_bio/tests/processes/__init__.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+**************
+
+ .. automodule:: server.tests.processor.base
```
Comment (tjanez) [inline resolwe_bio/tests/processes/__init__.py]: Am I right that this is a left-over from our previous internal version? If so, please, just remove these 2 lines.

**Comment by tjanez** [comment]
LGTM

**Pull request merged**